algolia/algoliasearch-sfcc-b2c
# feat: atomic reindexing

This PR adds the atomic reindexing logic to the `sendCategories` and `sendChunkOrientedProductUpdates` jobs.
An atomic reindex does the following:
- Delete any of the temporaries indices
- Index all products or categories into the temporary indices
- Copy the settings of the production indices on the temporary indices
- Move the temporary indices to production

### Changes

- Added new methods to `algoliaIndexingAPI`:
  - `deleteIndex`
  - `copyIndexSettings`
  - `moveIndex`
  - `waitTask`, that continuously poll the Algolia tasks API until a task is processed, for a maximum of 5 min
- Created a `reindexHelper`, to perform the needed operations to Algolia based on the indexType (products or categories) and the available locales:
  - delete the temporary indices
  - copy the settings
  - move the temporary indices
- Added `indexingMode` parameter to the sendProducts step, and use it to trigger the logic presented above
- Added `AlgoliaProductExport_v2` job definition

Methods of the `reindexHelper` will throw in case of error, so the jobs will be aborted before any data is lost (production indices are only overwritten at the very end).

### How to test :test_tube: 

I've added unit tests and updated the existing ones.

To test manually:
- Import the jobs definition
- Delete your Algolia indices
- Trigger the AlgoliaProductExport_v2 with `indexingMode=fullRecordUpdate`. You will see the indices being created directly (no `.tmp` index
- Trigger the AlgoliaProductExport_v2 again with `indexingMode=fullCatalogReindex`. You will see temporary indices being created during the process, and then moved to the production indices

---
SFCC-116
